### PR TITLE
Add erfcx with stable asymptotic expansion

### PR DESCRIPTION
**Pull request by TheSkyentist**
Adds jax.scipy.special.erfcx, the scaled complementary error function erfcx(x) = exp(x²) ·
erfc(x), addressing #1987.

### Implementation

The naive formula exp(x²) * erfc(x) overflows in float32 for x ≳ 9.4 and float64 for x ≳ 26.6,
even though erfcx(x) itself remains finite (it approaches 1/(√π·x) as x → ∞). To handle this,
the implementation is piecewise:

- Moderate x: exp(x²) * erfc(x) directly — no accuracy issues here since neither factor is
extreme
- Large positive x: asymptotic expansion (1/√π·x) · P(1/x²) via Horner, where P truncates the
series Σ (-1)^k (2k-1)!! / (2x²)^k — 5 terms for float32, 9 terms for float64

The switchover threshold is computed from dtypes.finfo(dtype).max as sqrt(log(fmax)), consistent
 with how igamma and similar functions handle overflow thresholds in JAX.

To prevent NaN propagation through the dead branch under JIT (the standard JAX pitfall with
jnp.where/lax.select), safe_x is clamped to 1.0 before the direct branch computation so exp(x²)
never actually overflows.

float16 and bfloat16 inputs are upcast to float32 before computation, following the bessel_i0e
pattern.

The JVP d/dx erfcx(x) = 2x·erfcx(x) − 2/√π is registered via custom_jvp/defjvps, consistent with
 how logit and xlogy are handled.

### AI Disclosure

Claude Code was used to aide in the code generation for this PR.

### CODE REVIEW

File: jax/_src/scipy/special.py [hunk 1]
```diff
@@ -578,6 +578,71 @@ def erfinv(x: ArrayLike) -> Array:
...
+    array containing values of the scaled complementary error function.
+
+  See also:
+    - :func:`jax.scipy.special.erfc`
+    - :func:`jax.scipy.special.erf`
+  """
+  x, = promote_args_inexact("erfcx", x)
+  return _erfcx(x)
+
+erfcx.defjvps(
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Let's put the `custom_jvp` on the `_erfcx` implementation instead: it pre-supposes that the inputs are floating point.

File: jax/_src/scipy/special.py [hunk 2]
```diff
@@ -578,6 +578,71 @@ def erfinv(x: ArrayLike) -> Array:
...
+    lambda g, ans, x: g * (2 * x * ans - _lax_const(x, 2. / np.sqrt(np.pi))))
+
+
+def _erfcx(x: Array) -> Array:
+  dtype = lax.dtype(x)
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Since `x` is an `Array` here, you can use `x.dtype` directly

File: jax/_src/scipy/special.py [hunk 3]
```diff
@@ -578,6 +578,71 @@ def erfinv(x: ArrayLike) -> Array:
...
+  if dtype == np.float64:
+    # At threshold ~26.6, first omitted term |c_9|/x^18 ~ 1e-21 << eps64 ~ 2e-16.
+    return _erfcx_impl(x, nterms=9)
+  elif dtype == np.float32:
+    # At threshold ~9.4, first omitted term |c_5|/x^10 ~ 5e-9 << eps32 ~ 1e-7.
+    return _erfcx_impl(x, nterms=5)
+  else:  # float16, bfloat16 — upcast to float32
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:619]: Complex input would run into this condition.

If this is not meant to support complex, then that should be checked somewhere.

File: jax/_src/scipy/special.py [hunk 4]
```diff
@@ -578,6 +578,71 @@ def erfinv(x: ArrayLike) -> Array:
...
+
+def _erfcx_asymptotic(x: Array, nterms: int) -> Array:
+  # Asymptotic expansion: erfcx(x) ~ (1/(sqrt(pi)*x)) * P(1/x^2)
+  # P(t) = sum_{k=0}^{N} c_k * t^k,  c_k = (-1)^k * (2k-1)!! / 2^k
+  _coeffs = [1., -.5, .75, -1.875, 6.5625, -29.53125, 162.421875,
+             -1055.7421875, 7918.06640625]
+  t = _lax_const(x, 1.) / lax.square(x)
+  p = _lax_const(x, _coeffs[nterms - 1])
+  for k in range(nterms - 2, -1, -1):
+    p = _lax_const(x, _coeffs[k]) + t * p
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: This looks like what `jnp.polyval` does. We should implement this using `polyval` (as a few other implementations in this file do) because it contains logic for rolling these loops to avoid long compile times.

File: tests/lax_scipy_special_functions_test.py [hunk 5]
```diff
@@ -231,6 +234,19 @@ def partial_lax_op(*vals):
...
                       rtol=.1, eps=1e-3)
 
+  def testErfcxLargeX(self):
+    # Verify no overflow and agreement with scipy in the asymptotic regime
+    # (float32: x > ~9.4, float64: x > ~26.6 — where exp(x^2) would overflow naively)
+    for x in [10., 20., 50., 100.]:
+      jax_val = float(lsp_special.erfcx(jnp.float32(x)))
+      scipy_val = float(osp_special.erfcx(np.float32(x)))
+      self.assertAllClose(jax_val, scipy_val, rtol=1e-5)
+    if np.float64 in jtu.dtypes.floating:
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py]: We should check this more directly: `if jax.config.x64_enabled`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great, thanks! A couple comments below.

Also, please add this new function to the docs at https://github.com/jax-ml/jax/blob/4182dd052b0b12028583d0c7b117d2118b7245ff/docs/jax.scipy.rst?plain=1#L155

**Comment by TheSkyentist** [comment]
Re: complex input support — for now this only supports real arguments, matching the existing convention for other functions in `jax.scipy.special` that reject complex inputs (`gammaln`, `beta`, `betainc`, `entr`, etc.).

Looked into what supporting complex would take: scipy's `erfcx` is computed via the MIT Faddeeva library (per the [0.12.0 release notes](https://docs.scipy.org/doc/scipy/release/0.12.0-notes.html)), using the identity `erfcx(z) = w(iz)` where `w` is the Faddeeva function. That's a fundamentally different algorithm from the one in this PR — the `1/x²` asymptotic series here is only valid in the sector `|arg(z)| < 3π/4`, and `lax.erfc` is real-only so the small-argument path doesn't generalize either. Full complex support would mean porting `wofz` (Faddeeva's `w(z)`) into JAX, which seems out of scope here. Happy to do that in a follow-up if useful.

**Comment by jakevdp** [comment]
Sounds good, I think leaving complex support out is fine.

File: jax/_src/scipy/special.py [hunk 6]
```diff
@@ -578,6 +578,71 @@ def erfinv(x: ArrayLike) -> Array:
...
+  formula which overflows.
+
+  Args:
+    x: arraylike, real-valued.
+
+  Returns:
+    array containing values of the scaled complementary error function.
+
+  See also:
+    - :func:`jax.scipy.special.erfc`
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:602]: We could add `erfcx` to the "See Also" sections of these APIs.

File: tests/lax_scipy_special_functions_test.py [hunk 7]
```diff
@@ -231,6 +234,19 @@ def partial_lax_op(*vals):
                       atol=.1 if jtu.test_device_matches(["tpu"]) else 1e-3,
                       rtol=.1, eps=1e-3)
 
+  def testErfcxLargeX(self):
+    # Verify no overflow and agreement with scipy in the asymptotic regime
+    # (float32: x > ~9.4, float64: x > ~26.6 — where exp(x^2) would overflow naively)
+    for x in [10., 20., 50., 100.]:
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py]: Let's test the full array in one call:
```
  x = np.array([10., 20., 50., 100.], dtype=np.float32)
  jax_val = lsp_special.erfcx(x)
  scipy_val = float(osp_special.erfcx(x))
  self.assertAllClose(jax_val, scipy_val, rtol=1e-5)
```
and similar below.

**Comment by jakevdp** [comment]
Also – once you've addressed the last couple comments, please squash all changes into a single commit (see https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests)

Thanks!

**Comment by TheSkyentist** [comment]
Thank you for the feedback! Squashed to one commit!

**Review by jakevdp — APPROVED** [review]
